switchlink/switchlink_address_test.cc

```diff
@@ -0,0 +1,324 @@
+                             switchlink_handle_t ecmp_h,
+                             switchlink_handle_t intf_h) {
+  struct test_results temp = {};
+  if (addr) {
+    temp.addr = *addr;
+  }
+  if (gateway) {
+    temp.gateway = *gateway;
+  }
+  if (intf_h) {
+    temp.intf_h = intf_h;
+  }
+  temp.opType = ADD_ADDRESS;
+  temp.num_handler_calls++;
+  results.push_back(temp);
+}
+
+/*
+ * Dummy function for switchlink_delete_route(). This function is
+ * invoked by switchlink_process_address_msg() when the msgtype is
+ * RTM_DELADDR for both IPv4 and IPv6 type of addresses. The actual
+ * method deletes route and removes entry from the database. Since this
+ * is dummy method, here the objective is to validate the invocation of
+ * this method with correct arguments. All the input params are stored
+ * in the test results structure and validated against each test case.
+ */
+
+void switchlink_delete_route(switchlink_handle_t vrf_h,
+                             const switchlink_ip_addr_t *addr) {
+  struct test_results temp = {};
+  if (addr) {
+    temp.addr = *addr;
+  }
+  temp.opType = DELETE_ADDRESS;
+  temp.num_handler_calls++;
+  results.push_back(temp);
+}
+
+/*
+ * Dummy function for switchlink_db_get_interface_info(). This function
+ * is invoked by switchlink_process_address_msg() for getting the intf
+ * info from the database. Since this is a dummy method, we are passing
+ * an ifindex 1 to this method and expects it to return an intf_info
+ * successfully with ifhandle being 0x10001.
+ */
+
+switchlink_db_status_t
+switchlink_db_get_interface_info(uint32_t ifindex,
+                                 switchlink_db_interface_info_t *intf_info) {
+  if (ifindex == 1) {
+    intf_info->intf_h = 0x10001;
+  }
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Test fixture.
+ */
+
+class SwitchlinkAddressTest : public ::testing::Test {
+protected:
+  struct nl_msg *nlmsg_ = nullptr;
+
+  // Sets up the test fixture.
+  void SetUp() override { ResetVariables(); }
+
+  // Tears down the test fixture.
+  void TearDown() override {
+    if (nlmsg_) {
+      nlmsg_free(nlmsg_);
+      nlmsg_ = nullptr;
+    }
+  }
+
+  void ResetVariables() {
+    // result variables
+    memset(&results, 0, sizeof(results));
+  }
+};
+
+/*
+ * Creates an IPv4 route
+ *
+ * Validates the switchlink_process_address_msg(). It parses an
+ * RTM_NEWADDR message which contains an IPv4 address and invokes
+ * switchlink_create_route() with the correct attributes.
+ *
+ * We invoke the switchlink_create_route() 2 times, one with the
+ * subnet mask derived from IFA_ADDRESS and another one with /32
+ * prefix. Hence we expect 2 test_results here, and this is the
+ * reason why test_results has been taken as a vector of size 2.
+ */
+
+TEST_F(SwitchlinkAddressTest, addIpv4Address) {
+  struct ifaddrmsg hdr = {
+      .ifa_family = AF_INET,
+      .ifa_prefixlen = 24,
+      .ifa_index = 1,
+  };
+
+  int prefix_len = 0;
```

Comment:
I'd recommend defining `prefix_len` closer to the point at which it's used. See below.